Adding new Live Debugger doc (#28365)
DataDog/documentation: content/en/tracing/live_debugger/_index.md

  text: "Sensitive Data Scrubbing"
- link: "/dynamic_instrumentation/symdb/"
  tag: "Documentation"
  text: "Autocomplete and Search (Preview)"
- link: "/tracing/error_tracking/execution_replay"
  tag: "Documentation"
  text: "Execution Replay"
---

{{< beta-callout-private url="https://www.datadoghq.com/product-preview/live-debugger/" >}}
    Live Debugger is in Limited Preview. Request access to join the waiting list.
{{< /beta-callout-private >}}

## Overview

With Live Debugger, you can debug running applications in real time, without redeploying code or interrupting service. Powered by Datadog's [Dynamic Instrumentation][1], Live Debugger uses logpoints&mdash;auto-expiring, "non-breaking breakpoints"&mdash;to collect information from running applications without pausing execution. This makes it ideal for investigating issues in environments where traditional debugging methods aren't practical.

## Key capabilities

Live Debugger provides:

- **Real-time inspection** of variable states, method arguments, and execution paths in running code.
- **Non-invasive data collection** that captures debugging information without stopping applications or degrading performance.

> - **Code instrumentation** with logpoints that can be added anywhere in your code, including third-party libraries.

Reviewer: Don't we restrict third-party libraries?

Author: I've heard both things. @OmerRaviv @shatzi halp please.

Reviewer 2: We allow instrument 3rd party methods.

The code origin and distributed debugger will not auto capture 3rd party code

- **Auto-expiring logpoints** that are automatically removed after a set time (default: 48 hours).
- **Conditional logging** based on user-defined criteria to capture data only when specific conditions are met.
- **Built-in [sensitive data scrubbing][3]** to prevent exposure of personal information, passwords, and secrets.

## Getting started

### Prerequisites

1. All [Dynamic Instrumentation prerequisites][16] are met.

> 2. (For one-click enablement) You have the following permissions: **Org Management**, **APM Remote Configuration Write**.

Author: @shatzi @etayyosefdd Can you guys please confirm these permissions are correct?

Reviewer 3: Also APM Remote Configuration Read

### Prerequisites

1. All [Dynamic Instrumentation prerequisites][16] are met.
1. You have [created a logs index][19] to store debugging information. 
1. (Recommended) You have enabled [Source Code Integration][20] to view and select specific code locations when adding logpoints.

### Setup Live Debugger

Enable and disable Live Debugger on a service using one of the following methods:

#### One-click enablement (recommended) ####

> <div class="alert alert-info">Only users with the following permissions can use one-click enablement: **Org Management**, **APM Remote Configuration Read**, **APM Remote Configuration Write**.</div>

Author: ```suggestion
<div class="alert alert-info">Only users with the following permissions can use one-click enablement: <b>Org Management, APM Remote Configuration Read, APM Remote Configuration Write</b>.</div>
```

**Performance impact**: Datadog's agent-driven instrumentation ensures minimal impact on application performance; sampling logic, rate limits, and built-in budgets prevent runaway data capture. The system limits logs with variable capture to 1 per second, but permits logs _without_ variable capture a higher sampling rate (due to less overhead).

**Pricing impact**: Logs captured by Datadog are all billed the same way, whether they are generated from Live Debugger or logger lines in your source code. With Live Debugger, the logpoints automatically expire after the set time period, limiting unnecessary data accumulation and costs. Monitor your [Datadog Plan & Usage page][21] for any unexpected increases after utilizing a new feature.

## Using Live Debugger

### Creating and using a Debug Session

Debug Sessions let you inspect your code at runtime with auto-expiring logpoints. To create and use a Debug Session:

1. Start a Debug Session from one of the following:
    - On the [Live Debugger page][14], click **Create Debug Session**.

>     - (Requires Code Origins feature) In the [Trace Explorer][22], click on a trace to open the side panel, find the Code Origin section, and click **Start Debug Session**.

Reviewer 4: Can we include a screenshot here to better direct folks?

Author: Will add these after pending UX changes!

1. Go to the [Live Debugger page][14].
2. Click **Create Debug Session**.
3. Choose your service, environment, and select where in your code to place the first logpoint.
4. Define a logpoint message template using the [Dynamic Instrumentation expression language][2].
5. (Optional) Enable "Capture Variables" to collect all execution context (this feature is rate-limited to 1 execution per second).
6. (Optional) Define a condition for when the logs should be emitted.

**Note:** Some feature limitations may apply depending on the service's runtime language. Review the [runtime language-specific documentation][17] for more details.

### Protecting sensitive data

Live Debugger data might contain sensitive information, especially when using the "Capture Variables" option. To protect this data:

1. Use Dynamic Instrumentation's built-in [sensitive data scrubbing][3] mechanisms.
2. Use [Sensitive Data Scanner][15] to identify and redact sensitive information based on regular expressions.

## Limitations

The following constraints apply to Live Debugger usage and configuration:

> - **Language support:** Live Debugger is available for the same runtime languages as [Dynamic Instrumentation][1].

Reviewer: I've seen hints at this elsewhere but it might be helpful to call out the various coverage across tracers.  not all libraries are necessarily supported in a particular tracer.

Author: For the time being, I'd rather add this info into the DI docs pages so we have a source of truth for support/compatibility. I can work on that in a separate PR.

- **Configuration scope:** Live Debugger and Dynamic Instrumentation are enabled or disabled together for the same service and environment.
- **Rate limits:**
   - Logpoints with variable capture: Limited to 1 execution per second.
   - Logpoints without variable capture: Limited to 5000 executions per second, per service instance.

> - **Session duration:** Debug Sessions automatically expire after 48 hours. This limit is not currently configurable.

Reviewer 4: ```suggestion
- **Session duration:** Debug Sessions automatically expire after 48 hours. This limit is not configurable.
```

1. Click "Enable" or "Disable":
    - "Enable" to allow users to create Debug Sessions on the selected service and environment.
    - "Disable" to deactivate active Debug Sessions and prevent users from creating more.

**Note**: No service restart is required for changes to take effect. Admins and security contacts receive email notifications when services are enabled or disabled.

#### Manual enablement ####
1. Select the service and environment on the [Live Debugger Settings][18] page.
1. Follow the instructions to enable Live Debugger.
1. Restart the service before using Live Debugger.

## Live Debugger and Dynamic Instrumentation
Due to shared underlying technology, Live Debugger and Dynamic Instrumentation are always enabled or disabled together on the same service and environment. 

Like Live Debugger, Dynamic Instrumentation allows users to create logpoints (in addition to supporting other custom instrumentation like spans, span tags, and metrics). However, Live Debugger logpoints expire automatically after a set time period, while Dynamic Instrumentation logpoints remain active until manually deactivated.

> When you enable or disable Live Debugger, the same action applies to Dynamic Instrumentation for that service and environment:

Author: ```suggestion
When you enable or disable Live Debugger, the same action applies to Dynamic Instrumentation for that service and environment. When disabled, all data capture stops from both active Debug Session logpoints and dynamic instrumentations.
```

- When enabled, users with appropriate permissions can create both Debug Sessions in Live Debugger and instrumentations in Dynamic Instrumentation.

> - When disabled, all data capture stops from both active Debug Session logpoints and dynamic instrumentations.

Author: ```suggestion
```

1. Use the built-in [sensitive data scrubbing][3] mechanisms.
2. Use [Sensitive Data Scanner][15] to identify and redact sensitive information based on regular expressions.

## Impact on performance and billing

Enabling Live Debugger and Dynamic Instrumentation on a service does not trigger data capture or impact performance. Data capture only occurs when there are active Debug Sessions or dynamic instrumentations on that service.

**Performance impact**: Datadog's agent-driven instrumentation ensures minimal impact on application performance; sampling logic, rate limits, and built-in budgets prevent runaway data capture.

**Pricing impact**: Logs captured by Datadog are all billed the same way, whether they are generated from Live Debugger or logger lines in your source code. With Live Debugger, the logpoints automatically expire after the set time period, limiting unnecessary data accumulation and costs. Monitor your [Datadog Plan & Usage page][21] for any unexpected increases after utilizing a new feature.

## Limitations

The following constraints apply to Live Debugger usage and configuration:

> - **Language support:** Live Debugger is available for the same runtime languages as [Dynamic Instrumentation][1].

Author: ```suggestion
- **Language support:** Live Debugger is available for the same runtime languages as [Dynamic Instrumentation][1], including: Java, Python, .NET, PHP (preview), Node.js (preview), Ruby (preview).
```

---
title: Live Debugger
further_reading:
- link: "/dynamic_instrumentation/"
  tag: "Documentation"
  text: "Dynamic Instrumentation"
- link: "/dynamic_instrumentation/expression-language/"
  tag: "Documentation"
  text: "Dynamic Instrumentation Expression Language"
- link: "/dynamic_instrumentation/sensitive-data-scrubbing/"
  tag: "Documentation"
  text: "Sensitive Data Scrubbing"
- link: "/dynamic_instrumentation/symdb/"
  tag: "Documentation"
  text: "Autocomplete and Search (Preview)"
- link: "/tracing/error_tracking/execution_replay"
  tag: "Documentation"
  text: "Execution Replay"
---

{{< beta-callout-private url="https://www.datadoghq.com/product-preview/live-debugger/" >}}
    Live Debugger is in Limited Preview. Request access to join the waiting list.
    <br>
    To submit questions, feedback, or requests related to Live Debugger, fill out <a href="https://docs.google.com/forms/d/e/1FAIpQLSdM9SV4fxrM_OvQ2CtI7CMl7evN0jasFb6X1QiPAbW6dPTQVQ/viewform?usp=header">this form</a> with details. 
    <br>

>     For urgent issues, contact <a href="https://www.datadoghq.com/support/">Datadog support</a>.

Reviewer 5: ```suggestion
```
This seems too scary?

Author: I just wanted to avoid people reporting time sensitive issues in the Google Form, just in case we don't see it fast enough. I'll change it to "time-sensitive" issues.

Author: ```suggestion
    For time-sensitive issues, contact <a href="https://www.datadoghq.com/support/">Datadog support</a>.
```

{{< /beta-callout-private >}}

## Overview

With Live Debugger, you can debug running applications in real time, without redeploying code or interrupting service. Powered by Datadog's [Dynamic Instrumentation][1], Live Debugger uses logpoints&mdash;auto-expiring, "non-breaking breakpoints"&mdash;to collect information from running applications without pausing execution. This makes it ideal for investigating issues in environments where traditional debugging methods aren't practical.

## Key capabilities

Live Debugger provides:

- **Real-time inspection** of variable states, method arguments, and execution paths in running code.
- **Non-invasive data collection** that captures debugging information without stopping applications or degrading performance.
- **Code instrumentation** with logpoints that can be added anywhere in your code, including third-party libraries.

> - **Auto-expiring logpoints** that are automatically removed after a set time (default: 48 hours).

Author: ```suggestion
- **Auto-expiring logpoints** that are automatically deactivated after a set time (default: 48 hours).
```

- **Conditional logging** based on user-defined criteria to capture data only when specific conditions are met.
- **Built-in [sensitive data scrubbing][3]** to prevent exposure of personal information, passwords, and secrets.

## Getting started

### Prerequisites

1. All [Dynamic Instrumentation prerequisites][16] are met.

> 2. (For one-click enablement) You have the following permissions: **Org Management**, **APM Remote Configuration Write**.

Author: ```suggestion
2. (For one-click enablement) You have the following permissions: **Org Management**, **APM Remote Configuration Read**, **APM Remote Configuration Write**.
```

> 2. (For one-click enablement) You have the following permissions: **Org Management**, **APM Remote Configuration Read**, **APM Remote Configuration Write**.

Reviewer 5: This isn't a prerequisite to use the product - should we move this to the "Setup Live Debugger" section? Seems like a more appropriate place for it.

> 3. You have [created a logs index][19] to store debugging information. 

Author: ```suggestion
1. You have [created a logs index][19] to store debugging information. 
```

> 4. (Recommended) You have enabled [Source Code Integration][20] to view and select specific code locations when adding logpoints.

Author: ```suggestion
1. (Recommended) You have enabled [Source Code Integration][20] to view and select specific code locations when adding logpoints.
```

### Setup Live Debugger

Enable and disable Live Debugger on a service using one of the following methods:

#### One-click enablement (recommended) ####

> <div class="alert alert-info">Only users with eligible permissions can use one-click enablement. Users may need to re-authenticate before using this feature.</div>

Reviewer 5: 1. As above, recommending that we move the blurb above about which permissions are needed here. 
2. Mentioning the need for re-authentication as a note above makes it sound like a warning/deterrent and might scare folks off, perhaps it's better to include it in the numbered list below instead?

1. Select the service and environment on the [Live Debugger Settings][18] page.

> 1. Check that all prerequisites are met (unmet requirements display on the Settings page).

Reviewer 5: ```suggestion
1. Check that all prerequisites are met as indicated on the page
```
(it's the same page, so weird to mention it twice I guess?)

Author: I wanted to distinguish between the Prerequisites section in the doc above vs. on the Settings page! I'll compromise ;) 

```suggestion
1. Check that all prerequisites are met as indicated on the Settings page.
```

1. Click "Enable" or "Disable":
    - "Enable" to allow users to create Debug Sessions on the selected service and environment.

>     - "Disable" to deactivate active Debug Sessions and prevent users from creating more.

Reviewer 5: This seems a bit redundant/stating the obvious. Perhaps it's enough to just mention that disabling will pause all active sessions?

Author: I don't think it's overkill to have both statements! I don't mind stating the obvious, just in case it's not obvious to the customer for whatever reason.

**Note**: No service restart is required for changes to take effect. Admins and security contacts receive email notifications when services are enabled or disabled.

#### Manual enablement ####
1. Select the service and environment on the [Live Debugger Settings][18] page.
1. Check that all prerequisites are met.
1. Follow the instructions to enable Live Debugger.
1. Restart the service before using Live Debugger.

> 

Reviewer 5: This seems a bit strange in that this is just describing a subset of what the in-app onboarding page will tell them to do. Maybe it's better to either:
1. Just tell them to go to the in-app onboarding page and follow the instructions
2. Send them to the Dynamic Instrumentation onboarding page, which has the full information on how to do set it up manually per language

I tend to prefer the former, because the in-app onboarding experience is a better experience to the static docs site?

Author: I could compromise and remove the second instruction to check the prerequisites!

I did want the opportunity to mention here in the docs that a service restart is required for this method.

Author: ```suggestion
#### Manual enablement ####
1. Select the service and environment on the [Live Debugger Settings][18] page.
1. Follow the instructions to enable Live Debugger.
1. Restart the service before using Live Debugger.

```

## Live Debugger and Dynamic Instrumentation
Due to shared underlying technology, Live Debugger and Dynamic Instrumentation are always enabled or disabled together on the same service and environment. 

Like Live Debugger, Dynamic Instrumentation allows users to create logpoints (in addition to supporting other custom instrumentation like spans, span tags, and metrics). However, Live Debugger logpoints expire automatically after a set time period, while Dynamic Instrumentation logpoints remain active until manually deactivated.

> When you enable or disable Live Debugger, the same action applies to Dynamic Instrumentation for that service and environment:
> - When enabled, users with appropriate permissions can create both Debug Sessions in Live Debugger and instrumentations in Dynamic Instrumentation.
> - When disabled, all data capture stops from both active Debug Session logpoints and dynamic instrumentations.

Reviewer 5: ```suggestion
```

Seems redundant, and also repeats what we already stated above?

Author: Again, yes a bit overkill, but at least there should be no misunderstandings about how DI will be impacted when they enable/disable Debugger (and vice-versa)! The point of this section is to clearly call out the relationship and cross-product impact of DI and Debugger - and to be able to link directly to this section when the question inevitably comes up (and not assume they have/will read the rest of the doc).

## Impact on performance and billing

Enabling Live Debugger and Dynamic Instrumentation on a service does not trigger data capture or impact performance. Data capture only occurs when there are active Debug Sessions or dynamic instrumentations on that service.

> **Performance impact**: Datadog's agent-driven instrumentation ensures minimal impact on application performance; sampling logic, rate limits, and built-in budgets prevent runaway data capture. The system limits logs with variable capture to 1 per second, but permits logs _without_ variable capture a higher sampling rate (due to less overhead).

Reviewer 5: ```suggestion
**Performance impact**: Datadog's agent-driven instrumentation ensures minimal impact on application performance; sampling logic, rate limits, and built-in budgets prevent runaway data capture. The system limits logs with variable capture to 1 per second, but permits logs _without_ variable capture a higher sampling rate (due to less overhead).
```

I'm not sure someone who's not familiar with the product would understand this point - maybe it's better to just explain this below, in the 'Creating logpoints' section?

Author: How about I move this to a subsection under "Using Live Debugger", below the "Protecting Sensitive Data" section? I can also just remove the sentence about the rate limit details since that's called out in a couple other places.

**Pricing impact**: Logs captured by Datadog are all billed the same way, whether they are generated from Live Debugger or logger lines in your source code. With Live Debugger, the logpoints automatically expire after the set time period, limiting unnecessary data accumulation and costs. Monitor your [Datadog Plan & Usage page][21] for any unexpected increases after utilizing a new feature.

## Using Live Debugger

### Creating and using a Debug Session

Debug Sessions let you inspect your code at runtime with auto-expiring logpoints. To create and use a Debug Session:

1. Start a Debug Session from one of the following:
    - On the [Live Debugger page][14], click **Create Debug Session**.

>     - (Requires Code Origins feature) In the [Trace Explorer][22], click on a trace to open the side panel, find the Code Origin section, and click **Start Debug Session**.

Author: I'm planning on adding screenshots to this section in a couple of weeks after some more UX changes for Live Debugger are completed, so I can add the gutter method to the list then as well!

We also discussed inserting the Create New Probe modal step for these CTAs instead of immediately starting a Debug Session, so that behavior might change soon.

### Creating logpoints

Logpoints are "non-breaking breakpoints" that specify where in the code to capture information, what data to include, and under what conditions. To add a logpoint for debugging:

1. Go to the [Live Debugger page][14].
2. Click **Create Debug Session**.
3. Choose your service, environment, and select where in your code to place the first logpoint.
4. Define a logpoint message template using the [Dynamic Instrumentation expression language][2].
5. (Optional) Enable "Capture Variables" to collect all execution context (this feature is rate-limited to 1 execution per second).
6. (Optional) Define a condition for when the logs should be emitted.

**Note:** Some feature limitations may apply depending on the service's runtime language. Review the [runtime language-specific documentation][17] for more details.

### Protecting sensitive data

Live Debugger data might contain sensitive information, especially when using the "Capture Variables" option. To protect this data:

1. Use Dynamic Instrumentation's built-in [sensitive data scrubbing][3] mechanisms.

> 2. Use [Sensitive Data Scanner][15] to identify and redact sensitive information based on regular expressions.

Reviewer 5: Both are mentioned on the first link, maybe just send them there? They'd need to follow the instructions on that page to use Sensitive Data Scanner successfully (see https://docs-staging.datadoghq.com/sarah.stonehill/live-debugger-docs/dynamic_instrumentation/sensitive-data-scrubbing/#customizing-sensitive-data-scanner)

Reviewer 5: and we can modify that page to refer to both Dynamic Instrumentation and Live Debugging?

## Limitations

The following constraints apply to Live Debugger usage and configuration:

> - **Language support:** Live Debugger is available for the same runtime languages as [Dynamic Instrumentation][1].

Reviewer 5: Better to just list them out here, to save the reader some time?

Author: Justin brought up that we should probably call out which frameworks/libraries are supported per runtime language as well. I'm inclined to add that info into a table under the Dynamic Instrumentation page tree, and link directly to the section from here. Would rather keep one source of truth if we can!

Author: Will leave as is!

- **Configuration scope:** Live Debugger and Dynamic Instrumentation are enabled or disabled together for the same service and environment.
- **Rate limits:**
   - Logpoints with variable capture: Limited to 1 execution per second.
   - Logpoints without variable capture: Limited to 5000 executions per second, per service instance.

> - **Session duration:** Debug Sessions automatically expire after 48 hours. This limit is not configurable.

Author: Hahaha this is exactly the word that @domalessi asked me to remove! 😂 

Here's a compromise, so that we don't HAVE to update it again after we implement the timeframe selector:

```suggestion
- **Session duration:** Debug Sessions automatically expire after 48 hours by default.
```
